### PR TITLE
Use MongoDB 💚 instead of Redis 💔

### PR DESCRIPTION
**Pull request by AbdullahAli**
MongoDB supports auto expiring documents - that is, documents which can auto-delete themselves after a specified number of seconds, please see https://docs.mongodb.com/manual/tutorial/expire-data for information

Using MongoDB in-conjunction with [Mongoose](https://mongoosejs.com/) significantly simplifies the code by removing the need for creating model-y type of utility functions for validations, etc (plusssss, [MongoDB](https://elements.heroku.com/addons/mongohq) is much cheaper to use than [Redis](https://elements.heroku.com/addons/openredis) on Heroku 💸

This PR more easily followed [commit-by-commit](https://github.com/Financial-Times/email-stethoscope/pull/12/commits)

That is all. Skrrr

giphy 6

### CODE REVIEW

File: config.js [hunk 1]
```diff
@@ -1,15 +1,10 @@
...
-	// event keys
-	eventKeyDelimiter: ':',
-	ageEventKeyPrefix: 'AGE',
+	dbURI: process.env.STETHOSCOPE_MONGO_URI,
+	eventTTL: process.env.STETHOSCOPE_EVENT_TTL || 86400
 };
 
 if (process.env.NODE_ENV !== 'production') {
-	config.redisUrl = 'redis://127.0.0.1:6379/0';
+	config.dbURI = 'mongodb://localhost:27017/ft-email-stethoscope-test';
```
Comment (andreaslof) [inline config.js]: Could we remove this `if`-statement and do like this instead:
```suggestion
	config.dbURI = process.env.STETHOSCOPE_MONGO_URI || 'mongodb://localhost:27017/ft-email-stethoscope-test';
```

Comment (AbdullahAli) [reply config.js]: Ah, good idea! I applied your suggested patch 👍

Comment (andreaslof) [reply config.js]: Ah, nifty! But could we move the `|| 'mongodb[…]'` default value to the config-block above? Sorry to be a bother!

File: lib/loggers/age.js [hunk 2]
```diff
@@ -1,35 +1,22 @@
...
-		if (!existingEvent.startDate) { return logger.warn(`Attempted to end an event that has not started.  key: '${key}'`); }
-		if (existingEvent.endDate) { return logger.warn(`Attempted to end an event that has already ended.  key: '${key}'`); }
+		const existingEvent = await AgeEvent.findOne({ event, identifier });
+		if (existingEvent.endedAt) { return logger.warn(`Attempted to end an event that has already ended. ${{ event, identifier }}`); }
 
-		return redis.hmset(key, { endDate: Date.now() });
+		return await existingEvent.set({ endedAt: Date.now() }).save();
 	} catch (err) {
-		return errorLogger.logUnexpectedError(err, key);
+		return logger.error(`Unexpected error occured: '${err.message}'`);
```
Comment (keirog) [inline lib/loggers/age.js]: _Super_ minor: ~occured~ occurred

File: lib/metrics/age.js [hunk 3]
```diff
@@ -1,48 +1,36 @@
...
-
 	try {
-		const existingEvent = await redis.hgetall(key);
+		const existingEvent = await AgeEvent.findOne({ event, identifier });
 
-		return (Object.keys(existingEvent).length) ? _age(existingEvent) : undefined;
+		return existingEvent ? existingEvent.age : undefined;
 	} catch (err) {
-		return errorLogger.logUnexpectedError(err, key);
+		return logger.error(`Unexpected error occured: '${err.message}'`);
```
Comment (keirog) [inline lib/metrics/age.js]: Super minor: ~occured~ occurred

File: lib/metrics/age.js [hunk 4]
```diff
@@ -1,48 +1,36 @@
...
-
 	try {
 		if (!validOperations[operation]) { return logger.error(`Attempted to use invalid operation: ${operation}.  Valid operations are: ${Object.keys(validOperations)}`); }
-		const existingEvents = await getAllEvents(`${key}*`);
 
+		const existingEvents = await AgeEvent.find({ event });
 		return existingEvents.length ? validOperations[operation](existingEvents) : undefined;
 	} catch (err) {
-		return errorLogger.logUnexpectedError(err, key);
+		return logger.error(`Unexpected error occured: '${err.message}'`);
```
Comment (keirog) [inline lib/metrics/age.js]: Super minor: ~occured~ occurred

File: lib/metrics/age.js [hunk 5]
```diff
@@ -1,48 +1,36 @@
...
 exports.orderedEvents = async ({ limit } = {}) => {
-	const key = eventKey({ prefix });
-
 	try {
-		const events = (await getAllEvents(`${key}*`)).map(decorateWithAge).sort(compareByAge).slice(0, limit);
+		const events = (await AgeEvent.find()).sort(compareByAge).slice(0, limit);
 		return events.length ? events : undefined;
 	} catch (err) {
-		return errorLogger.logUnexpectedError(err);
+		return logger.error(`Unexpected error occured: '${err.message}'`);
```
Comment (keirog) [inline lib/metrics/age.js]: Super minor: ~occured~ occurred

File: config.js (hunk 1 above)
Comment (AbdullahAli) [reply config.js]: Ah, my bad, it should be fixed now 🙇

File: test/lib/loggers/age.test.js [hunk 6]
```diff
@@ -88,62 +71,44 @@ describe('Loggers > Events Age', () => {
...
-				const errorLoggerStub = sinon.stub(errorLogger, 'logUnexpectedError');
+				await startEvent({ event, identifier });
+
+				sinon.stub(AgeEvent.prototype, 'save').rejects(expectedError);
+				const loggerStub = sinon.stub(logger, 'error');
 
 				await endEvent({ event, identifier });
 
-				expect(errorLoggerStub.calledWith(expectedError, expectedKey)).to.be.true;
+				expect(loggerStub.calledWith(`Unexpected error occured: '${expectedError.message}'`)).to.be.true;
```
Comment (keirog) [inline test/lib/loggers/age.test.js]: Super minor: ~occured~ occurred

File: lib/loggers/age.js [hunk 7]
```diff
@@ -1,35 +1,22 @@
...
 	try {
-		const existingEvent = await redis.exists(key);
-		if (existingEvent) { return logger.warn(`Attempted to start an event that has already been started.  key: ${key}`); }
-		await redis.hmset(key, { event, identifier, startDate: Date.now() });
-		return redis.expire(key, expire);
+		if (await AgeEvent.findOne({ event, identifier })) { return logger.warn(`Attempted to start an event that has already been started. ${{ event, identifier }}`); }
+		return await AgeEvent.create({ event, identifier, createdAt: Date.now() });
 	} catch (err) {
-		return errorLogger.logUnexpectedError(err, key);
+		return logger.error(`Unexpected error occured: '${err.message}'`);
```
Comment (keirog) [inline lib/loggers/age.js]: Super minor: ~occured~ occurred

File: test/lib/metrics/age.test.js [hunk 8]
```diff
@@ -48,15 +51,14 @@ describe('Metrics > Events Age', () => {
...
 
-				const errorLoggerStub = sinon.stub(errorLogger, 'logUnexpectedError');
-				sinon.stub(Redis.prototype, 'hgetall').throws(expectedError);
+				const loggerStub = sinon.stub(logger, 'error');
+				sinon.stub(AgeEvent, 'findOne').rejects(expectedError);
 
 				const age = await eventAge({ event, identifier });
 
-				expect(errorLoggerStub.calledWith(expectedError, expectedEventKey)).to.be.true;
+				expect(loggerStub.calledWith(`Unexpected error occured: '${expectedError.message}'`)).to.be.true;
```
Comment (keirog) [inline test/lib/metrics/age.test.js]: Super minor: ~occured~ occurred

File: test/lib/metrics/age.test.js [hunk 9]
```diff
@@ -150,22 +154,21 @@ describe('Metrics > Events Age', () => {
...
-				const errorLoggerStub = sinon.stub(errorLogger, 'logUnexpectedError');
-				sinon.stub(Redis.prototype, 'hgetall').throws(expectedError);
+				const loggerStub = sinon.stub(logger, 'error');
+				sinon.stub(AgeEvent, 'find').rejects(expectedError);
 
-				await eventAge({ event });
+				await eventsAge({ event, operation: 'max' });
 
-				expect(errorLoggerStub.calledWith(expectedError, expectedEventKey)).to.be.true;
+				expect(loggerStub.calledWith(`Unexpected error occured: '${expectedError.message}'`)).to.be.true;
```
Comment (keirog) [inline test/lib/metrics/age.test.js]: Super minor: ~occured~ occurred

File: test/lib/loggers/age.test.js [hunk 10]
```diff
@@ -1,84 +1,69 @@
...
 
-				sinon.stub(Redis.prototype, 'exists').throws(expectedError);
-				const errorLoggerStub = sinon.stub(errorLogger, 'logUnexpectedError');
+				sinon.stub(AgeEvent, 'create').rejects(expectedError);
+				const loggerStub = sinon.stub(logger, 'error');
 
 				await startEvent({ event, identifier });
 
-				expect(errorLoggerStub.calledWith(expectedError, expectedKey)).to.be.true;
+				expect(loggerStub.calledWith(`Unexpected error occured: '${expectedError.message}'`)).to.be.true;
```
Comment (keirog) [inline test/lib/loggers/age.test.js]: Super minor: ~occured~ occurred

File: lib/loggers/age.js (hunk 2 above)
Comment (AbdullahAli) [reply lib/loggers/age.js]: Ah, soz, fixed 👍

**Comment by AbdullahAli** [comment]
yay `close(force: Boolean)`

https://github.com/Financial-Times/email-stethoscope/pull/12/files#diff-9d893d8a174c0e3f0fcd8b25135abb1eR15

**Pull request merged**